Allow usage of Closures in yii\filters\PageCache for enabled and variations

C4RoCKeT opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Fixed issues  | #16649

C4RoCKeT committed:
closure support for 'variations' and 'enabled'

When using the yii\rest\activeController, I wanted to vary the cache based on the user. While defining the behaviors, the user identity is not yet available and thus, Yii::$app->user->id always returned null.

Because of this I've added support for Closures when checking `$this->enabled` in `beforeAction()` and for `$this->variations` in `calculateCacheKey`.

Examples:

`'enabled' => function (PageCache $pageCache) {
    return Yii::$app->request->isGet && !Yii::$app->user->identity->isAdmin();
},`

and:

`'variations' => [
    function () {
        return Yii::$app->user->id;
    },
    Yii::$app->request->getQueryParams()
]`

C4RoCKeT committed:
Update CHANGELOG.md

rob006 commented on framework/filters/PageCache.php:
```diff
@@ -155,7 +158,7 @@ public function init()
     public function beforeAction($action)
     {
-        if (!$this->enabled) {
+        if (($this->enabled instanceof \Closure && !call_user_func($this->enabled, $this)) || !$this->enabled) {
```
This looks incorrect. If closure will return `false `, it will be used again in `!$this->enabled` (which will always evaluate to `false`). Although it may work, it looks weird and relies on a bit of luck on type juggling.

C4RoCKeT replied:
`!call_user_func($this->enabled, $this)` so, if the closure returns `false`, it will become `true` in the statement and will not trigger the evaluation of the second statement, thus never checking `!$this->enabled`.

rob006 replied:
And what will happen if closure return `true`?

C4RoCKeT replied:
Ah snap, you are right. Didn't think that one through.

`if($this->enabled instanceof \Closure) {
    $this->enabled = call_user_func($this->enabled, $this);
}
if (!$this->enabled) {`

This probably would be a cleaner solution.

rob006 commented on framework/filters/PageCache.php:
```diff
@@ -298,6 +301,11 @@ protected function calculateCacheKey()
+            if ($variation instanceof \Closure) {
+                $variation = call_user_func($variation, $this);
+            }
+        }
```
I don't think that we should modify `$this->variations` directly. This will create non-obvious side effect.

C4RoCKeT committed:
Cleaner solution for the $this->enabled closure

Added an extra if statement, because one statement relied too much on type juggling.
